### ISSUE TITLE
Show current file feature

### ISSUE THREAD

**Issue by maxim** [divmain/GitSavvy#102]
Git package had a very convenient feature: "show current file". It's just like GitSavvy's "log current file", but instead of taking you to full commit diff, it just opens current file at chosen revision. It helps you see how a particular file looked in the past, and manually bring back some code. I don't see an equivalent in GitSavvy, would be awesome to have it. That's pretty much the only thing I'm missing from Git.

**Comment by divmain** [comment]
It shouldn't be difficult to implement as described.  I've also considered a view where you can walk back in time for a particular file.

For example, you open the palette, type something like `git: show file at...`, select a revision, and the view pops up.  This sounds like basically what you've described.  However, at that point, you could then hit maybe <kbd>Super</kbd>+<kbd>Left</kbd>, and the view would change to display the revision immediately before the visible one.

Thoughts @maxim?

**Comment by maxim** [comment]
@divmain This walking sounds super convenient (which is what I love about GitSavvy, you just keep taking things a step or two further). Imho, it's such a clean addition that it doesn't even have to be done in 1st iteration of this feature, can be added later, but yes, definitely would instause it.

**Comment by maxim** [comment]
I would give default key combo a bit more thought however. <kbd>Super</kbd>+<kbd>Left</kbd> and some other similar bindings are often taken, but say my <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>Left</kbd> is available.

**Comment by divmain** [comment]
Hmm, I think <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>Left</kbd> may effect the screen orientation in Windows 7 and later.  How about:

- <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>Left</kbd> - walk back in time (Windows)
- <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>Right</kbd> - walk forward in time (Windows)
- <kbd>Super</kbd>+<kbd>Shift</kbd>+<kbd>Left</kbd> - walk back in time (Linux/OS X)
- <kbd>Super</kbd>+<kbd>Shift</kbd>+<kbd>Right</kbd> - walk forward in time (Linux/OS X)

**Comment by maxim** [comment]
Problem is, you still want to select/copy/navigate the text in that file, and <kbd>Super</kbd>+<kbd>Left</kbd>/<kbd>Right</kbd> is used to jump to beginning/end of line. Holding <kbd>shift</kbd> does the same + selection. So the advantage of <kbd>Ctrl</kbd>+<kbd>Alt</kbd> is that at least it's doesn't seem to be bound to any standard editor functions, and these keys would still be configurable.

**Comment by divmain** [comment]
We need more modifier keys!  Maybe left/right keys aren't the right thing to use here, even though they're the most obvious.

**Comment by maxim** [comment]
Yeah, it's hard to pick the keys. Also it's complicated by the the fact that even though the file is non-editable, you still wouldn't want to restrict navigation in it, including for people like myself who use Vintage/Vintageous (vim simulation).

**Comment by alextegelid** [comment]
I second this, especially with the walking (next/prev) feature. 


How about this for keyboard shortcuts? Ctrl+super+alt as modifiers, and dot and comma as next/prev (or vice versa, whichever makes the most sense).


These keybindings would then be similar to the inline diff view where dot and comma is used to walk through the hunks.

**Comment by ethanbustad** [comment]
:+1: for the feature, :+1: for using dot/comma to mimic inline diff

**Comment by zx8** [comment]
Seems like an awesome idea! +1

**Comment by maxim** [comment]
I love dot/comma idea, but it's best to avoid all 3 modifiers, very inconvenient. Ideal for me would be <kbd>Ctrl</kbd>+<kbd>Super</kbd>+<kbd>,</kbd> (no alt)

**Comment by yyyc514** [comment]
> It helps you see how a particular file looked in the past, and manually bring back some code. 

What exactly would that workflow look like?

- pop open file at revision 1234
- page forward/backwards thru history
- copy stuff
- close tab
- now we're back to the editable file
- paste

Or something else?

**Comment by maxim** [comment]
@yyyc514 exactly. Also if you're refactoring something, and in the process lost something, this is a great UI to quickly refer to what the logic should've been from before refactor, so you can reproduce it differently in the refactored version.

**Comment by yyyc514** [comment]
What's the best way to do this and preserve history across renames?  `git log --follow` can follow the history for a file but `git show hash:file` doesn't work if the file name does not match the exact name of the file as of the hash you passed in.

**Comment by maxim** [comment]
That's a good question. I believe the old sublime Git package (which has this feature) doesn't follow files across renames. Would be cool if we could find a way to do it.

**Comment by maxim** [comment]
Another small detail that could be useful - when walking back-and-forth across revisions, perhaps we should try to preserve the scroll position. The code won't be where it was, and if the scroll is beyond the limit of last revision - we'd resort to just scrolling to the bottom, but imo it's better than not doing it.

**Comment by yyyc514** [comment]
That would be a neat trick but probably pretty hard to do without a lot of extra logic (stuff that's already built into things like git blame, etc).

**Comment by yyyc514** [comment]
A git blame walker could be a lot more powerful than a git log walker, but you'd be limited to jumping to commits that had some remnants existing in the current revision of the file.

**Comment by divmain** [comment]
@maxim, preserving scroll/cursor position should definitely be do-able and will be included.

@yyyc514, I've had the same thought about blame.  Once this history-walker feature lands, a keyboard shortcut could be added to the blame view to take you to the revision at that time.

In my ideal world, I would see an inline-diff highlighted file as I walked back in time, showing the parts of the file that changed between the revision I'm viewing and its parent.  But considering the complexity, I think that is over-reaching for a v1, and I'll probably spin that off into another issue to revisit some day.

**Comment by yyyc514** [comment]
> preserving scroll/cursor position is 100% do-able and will be included.

Neat, but how do you get that information without git blame?  A line could move 1000 lines between commits... I know git blame (the version meant to be parsed) has all that info but didn't think you would get that granular with just commits - without rebuilding all that line counting code yourself.

> In my ideal world, I would see an inline-diff highlighted file as I walked back in time, showing the parts of the file that changed between the revision I'm viewing and its parent.

That would be interesting.  What I typically want to do is walk a single line back thru it's history inside git blame... so I'm seeing the files as they stood and who did what.  I can more about what the file looks like at a single revision than the diff... and if I want the diff it's only a keystroke away.  I'll make an issue for what I have in mind.

**Comment by divmain** [comment]
From #276:

> I think it would be useful for chain like this:
> show commit > open file > show log of this file
> Open file works from DIFF view by 'o' shortkey but doesn't work from COMMIT view.

This sub-feature should be included as part of this effort.  For a walk through history, we'll need an `show_file_at_commit` command.  To "open file" as suggested by @dlnsk, we'll need this functionality - the file doesn't necessarily still exist at HEAD.

**Comment by divmain** [comment]
Note to self:
```
git log --reverse --ancestry-path COMMIT_BEING_VIEWED...HEAD
```

**Comment by dlnsk** [comment]
And another one feature request:
Will be cool if some shortcut (maybe Ctrl+R) on DIFF/COMMIT open popup with list of files changed in it with the ability to go to this part of diff.

**Comment by randy3k** [comment]
the `git: log current file` and `git: graph current file` feature on "master" branch provide a partial solution to this issue.

**Comment by reagle** [comment]
Was the idea of walking through the history and seeing the (w)diff's for a given file or the repo ever done? I can't figure out how to do it.
